lib/postgres.go

```diff
@@ -0,0 +1,1414 @@
+		}
+	}
+
+	if len(metadataBlockRewards) > 0 {
+		if _, err := tx.Model(&metadataBlockRewards).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataBitcoinExchanges) > 0 {
+		if _, err := tx.Model(&metadataBitcoinExchanges).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataPrivateMessages) > 0 {
+		if _, err := tx.Model(&metadataPrivateMessages).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataSubmitPosts) > 0 {
+		if _, err := tx.Model(&metadataSubmitPosts).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataUpdateProfiles) > 0 {
+		if _, err := tx.Model(&metadataUpdateProfiles).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataExchangeRates) > 0 {
+		if _, err := tx.Model(&metadataExchangeRates).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataFollows) > 0 {
+		if _, err := tx.Model(&metadataFollows).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataLikes) > 0 {
+		if _, err := tx.Model(&metadataLikes).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataCreatorCoins) > 0 {
+		if _, err := tx.Model(&metadataCreatorCoins).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataCreatorCoinTransfers) > 0 {
+		if _, err := tx.Model(&metadataCreatorCoinTransfers).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataSwapIdentities) > 0 {
+		if _, err := tx.Model(&metadataSwapIdentities).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (postgres *Postgres) UpsertBlockAndTransactions(blockNode *BlockNode, bitcloutBlock *MsgBitCloutBlock) error {
+	return postgres.db.RunInTransaction(postgres.db.Context(), func(tx *pg.Tx) error {
+		err := postgres.UpsertBlockTx(tx, blockNode)
+		if err != nil {
+			return err
+		}
+
+		blockHash := blockNode.Hash
+		err = postgres.UpsertChainTx(tx, "main", blockHash)
+		if err != nil {
+			return err
+		}
+
+		err = postgres.InsertTransactionsTx(tx, bitcloutBlock.Txns, blockHash)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (postgres *Postgres) GetUtxoEntryForUtxoKey(utxoKey *UtxoKey) *UtxoEntry {
+	utxo := &PGTransactionOutput{
+		OutputHash:  &utxoKey.TxID,
+		OutputIndex: utxoKey.Index,
+		Spent:       false,
+	}
+
+	err := postgres.db.Model(utxo).WherePK().Select()
+	if err != nil {
+		return nil
+	}
+
+	return &UtxoEntry{
+		PublicKey:   utxo.PublicKey,
+		AmountNanos: utxo.AmountNanos,
+		// TODO: Block height?
+		UtxoType: utxo.OutputType,
+		isSpent:  utxo.Spent,
+		UtxoKey:  utxoKey,
+	}
+}
+
+func (postgres *Postgres) GetUtxoEntriesForPublicKey(publicKey []byte) []*UtxoEntry {
+	var transactionOutputs []*PGTransactionOutput
+	err := postgres.db.Model(&transactionOutputs).Where("public_key = ?", publicKey).Select()
```

Comment:
I would maybe use fmt.Sprintf("%v = ?", PublicKeyString) in cases like these to avoid getting fucked when you change something. But it does make it pretty painful to read...

Reply:
Sigh.. on second thought it is a lot harder to read... I think instead maybe we'll just never ever change column names...